types/wlr_xdg_shell_v6.c

```diff
@@ -677,15 +663,38 @@ static void xdg_toplevel_protocol_unset_maximized(struct wl_client *client,
 static void xdg_toplevel_protocol_set_fullscreen(struct wl_client *client,
 		struct wl_resource *resource, struct wl_resource *output_resource) {
 	struct wlr_xdg_surface_v6 *surface = wl_resource_get_user_data(resource);
+
+	struct wlr_output *output = NULL;
+	if (output_resource != NULL) {
+		output = wl_resource_get_user_data(output_resource);
+	}
+
 	surface->toplevel_state->next.fullscreen = true;
-	wl_signal_emit(&surface->events.request_fullscreen, surface);
+
+	struct wlr_xdg_toplevel_v6_set_fullscreen_event event = {
+		.client = client,
+		.surface = surface,
+		.fullscreen = true,
+		.output = output,
+	};
+
+	wl_signal_emit(&surface->events.request_fullscreen, &event);
 }
 
 static void xdg_toplevel_protocol_unset_fullscreen(struct wl_client *client,
 		struct wl_resource *resource) {
 	struct wlr_xdg_surface_v6 *surface = wl_resource_get_user_data(resource);
+
 	surface->toplevel_state->next.fullscreen = false;
-	wl_signal_emit(&surface->events.request_fullscreen, surface);
+
+	struct wlr_xdg_toplevel_v6_set_fullscreen_event event = {
+		.client = client,
+		.surface = surface,
+		.fullscreen = false,
+		.output = NULL,
+	};
+
+	wl_signal_emit(&surface->events.request_fullscreen, &event);
```

Comment:
I think we wait until the shell surface commits the whole state before we emit the event.

Reply:
What do you mean? We should wait? Right now `request_fullscreen` isn't triggered on commit.

Note: the state doesn't keep track of the `output` parameter, which would make waiting for a commit more tricky.

Reply:
I'm making a note of this now because I'm pretty sure from looking at the code we're going to see rendering artifacts when the fullscreen happens. I'll test it when you make the command.